### ISSUE TITLE
Only send authentication header for the correct registries

### ISSUE THREAD

**Issue by buschtoens** [yarnpkg/yarn#3407]
**Do you want to request a *feature* or report a *bug*?**

Bug.

**What is the current behavior?**

If the user previously logged in to some registry, for instance the public npm registry, yarn always sends an `Authorization` header when requesting scoped packages, even if these are unrestricted packages or are requested from a different registry.

Some private registries (specifically [Nexus, issue #13145](https://issues.sonatype.org/browse/NEXUS-13145)) send a 401 response for these requests, as they can't verify the token (even though they actually don't need to).

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Log in to the public npm registry via `npm login`.

2. Set a different/private registry in the config file.

3. Install a scoped package, for example `yarn add @glimmer/resolver`.

4. Yarn sends the `Authorization` header.

**What is the expected behavior?**

Yarn should only send the authentication header for the registry the user actually authenticated against. Funnily enough, yarn kind of has a notion of which registry the user authenticated against, but it's variable.

```
$ npm login
$ yarn config list
yarn config v0.23.4
info yarn config
{ 'version-tag-prefix': 'v',
 ...
  'user-agent': 'yarn/0.23.4 npm/? node/v7.10.0 linux x64',
  email: 'buschtoens@gmail.com',
  lastUpdateCheck: 1494829799781,
  username: 'buschtoens' }
info npm config
{ '//registry.npmjs.org/:_authToken': '***' }

$ echo "registry=http://my-private-registry.com/" > .npmrc
$ yarn config list
yarn config v0.23.4
info yarn config
{ 'version-tag-prefix': 'v',
  ...
  'user-agent': 'yarn/0.23.4 npm/? node/v7.10.0 linux x64',
  email: 'buschtoens@gmail.com',
  lastUpdateCheck: 1494829799781,
  username: 'buschtoens' }
info npm config
{ '//registry.npmjs.org/:_authToken': '***',
  registry: 'http://my-private-registry.com/' }
```

**Please mention your node.js, yarn and operating system version.**

**Comment by bestander** [comment]
A PR is welcome

**Comment by ErisDS** [comment]
For anyone else who runs into this, based on #3407 I've managed to [figure out a workaround](https://github.com/yarnpkg/yarn/issues/3093#issuecomment-306413909), by commenting several lines of my `~/.npmrc` file.

**Comment by vicapow** [comment]
I sent an email to github because I believe they may have changed their behavior around authentication recently. Folks have reported that this issue is impacting all versions of yarn and npm v5 (not v4) when using a private registry and npm modules that reference git SHAs (probably infrequent possibility for most folks). As a workaround, use npm v4.

````
Hello,

I noticed that github is returning 404 Not Found when providing an invalid
"authorization" header for normally publicly accessible resources.

Not entirely sure if this is a bug but this behavior affects folks who
configure their npm or yarn to provide authentication headers
https://github.com/yarnpkg/yarn/issues/3407 for private registries
because npm and yarn provide the "authorization" header for all requests.

Here's two simple curl requests to demonstrate the issue:

expected 200 OK, actual 200 OK

(request without "authorization" header)

curl -H 'Accept: application/octet-stream' -H 'Accept-Encoding: gzip' -H
'User-Agent: unpm-allow-external-deprecated-packages'
https://codeload.github.com/mapbox/mapbox-gl-style-spec/tar.gz/83b1a3e5837d785af582efd5ed1a212f2df6a4ae
--verbose -o /dev/null

expected 200 OK, actual 404 Not Found

(request with "authorization" header)

curl -H 'authorization: Basic XXXXX-unpm-token-XXXXXXXX' -H 'Accept:
application/octet-stream' -H 'Accept-Encoding: gzip' -H 'User-Agent:
unpm-allow-external-deprecated-packages'
https://codeload.github.com/mapbox/mapbox-gl-style-spec/tar.gz/83b1a3e5837d785af582efd5ed1a212f2df6a4ae
--verbose -o /dev/null

All the best,
Victor
````

**Comment by BYK** [comment]
This is fixed on master thanks to #3987.
